### ISSUE TITLE
TFLite micro: Is the flatbuffer allowed to be placed in flash in all cases?

### ISSUE THREAD

**Issue by janjongboom** [tensorflow/tensorflow#41271]
@tensorflow/micro

**System information**

- TensorFlow installed from (source or binary):

Source.

- Tensorflow version (commit SHA if source):

b2f092894012e9c2612cb46c332140f28d91ced2

- Target platform (e.g. Arm Mbed OS, Arduino Nano 33 etc.):

Arduino Nano 33 BLE Sense

**Describe the problem**

We received a bug report from a user that is experiencing a hard fault when initializing their TFLite model when also adding the ArduinoBLE library on the Arduino Nano 33 BLE Sense. So far we have always placed the flatbuffer model in flash, and this has not been a problem - but when digging into this bug report I traced the error back to the first call to the `InitializeTfLiteTensorFromFlatbuffer`. Is it possible that the flatbuffer is manipulated here? When moving the flatbuffer byte array into RAM the problem disappears. It's hard to debug on my end as I haven't figured out how to connect my J-Link to this board and build with debug symbols enabled, and because the way the serial communications are handled I can't see the error log either.

**Please provide the exact sequence of commands/steps when you ran into the problem**

I've reproduced the problem with:

* Arduino Nano 33 BLE Sense
* Arduino Core for nRF528x (Mbed OS) 1.1.4
* Arduino BLE 1.1.3
* The attached library (add via **Sketch > Include library > From ZIP file**)

And the following sketch:

```cpp
#include <ArduinoBLE.h>
#include <continuous_gestures_inference.h>

BLEService soundsService("180C");  // User defined service

static const float features[] = {
19.9700, 8.7300, 17.5700, 19.9700, 8.7300, 17.5700, 19.9700, 9.8200, 14.0300, 15.9400, 10.0700, 14.1400, 5.1400, 10.3500, 14.0700, -3.6600, 8.7700, 13.4100, -14.0000, 4.9900, 8.9200, -19.9800, -1.5200, 5.5700, -19.9800, -1.5200, 5.5700, -19.9800, -8.7800, -7.7400, -19.9800, -8.9900, -19.5300, -19.9800, -4.7900, -6.8400, -19.9800, 1.8400, 9.2000, -15.4000, 9.3800, 15.2200, -7.5900, 9.4600, 19.9500, -7.5900, 9.4600, 19.9500, 0.3900, 14.4600, 19.9800, 16.8500, 19.5300, 19.9800, 19.9700, 18.9900, 19.9800, 19.9700, 19.8700, 19.9500, 19.9700, 19.9700, 18.2000, 19.9700, 19.9700, 16.2600, 19.9700, 19.9700, 16.2600, 19.9700, 16.1000, 17.7500, 19.9700, 8.9900, 17.6500, 15.7300, 3.6000, 15.1800, 11.4500, -0.7100, 17.0700, 7.3700, -3.4600, 19.4300, 7.3700, -3.4600, 19.9800, 1.4400, -4.7400, 19.9800, -5.8700, -4.6500, 18.4300, -17.6300, -2.4500, 13.9000, -19.9800, -2.8000, 12.0600, -19.9800, 0.3900, 6.8400, -19.9800, 5.3700, -12.8800, -19.9800, 5.3700, -12.8800, -19.9800, 6.8900, -19.8900, -19.9800, 3.8400, -13.2000, -19.9800, 0.6700, -8.7200, -19.9800, 0.3600, -8.4200, -19.9800, 0.1200, -4.4400, -19.9800, 0.1300, 5.5800, -19.9800, 0.1300, 5.5800, -14.2500, -7.5100, 19.4800, -4.9700, -7.8700, 19.9800, 7.2300, -5.0500, 19.9800, 19.8600, 5.9300, 19.9800, 19.9700, 10.6000, 19.9800, 19.9700, 7.5400, 19.9800, 19.9700, 7.5400, 19.9800, 19.9700, 10.1600, 19.9800, 19.9700, 10.1500, 19.7000, 19.9700, 6.8200, 16.7800, 19.8700, 3.9500, 15.8000, 12.7500, 0.4500, 5.5200, 6.5800, -1.7100, 12.2200, 6.5800, -1.7100, 12.2200, -0.4100, -3.4200, 11.1400, -10.6100, -0.6800, 4.3500, -19.1800, -0.6900, 1.0500, -19.9800, -2.2000, -2.9700, -19.9800, -3.8700, -12.7700, -19.9800, -3.8700, -12.7700, -19.9800, -2.6600, -15.4700, -19.9800, 0.3700, -5.8800, -19.9800, 5.4000, 3.2500, -18.2600, 9.8800, 11.9400, -11.9500, 9.2100, 19.5900, -5.1000, 11.2400, 19.9800, -5.1000, 11.2400, 19.9800, 8.0700, 15.0100, 19.9800, 18.6900, 16.8800, 19.9800, 19.9700, 15.9200, 19.9800, 19.9700, 15.8600, 19.9800, 19.9700, 16.9700, 19.9800, 19.9700, 16.3800, 19.9800, 19.9700, 16.3800, 19.9800, 19.9700, 14.5000, 19.9800, 19.9700, 11.1800, 19.9800, 17.8900, 5.6400, 19.9800, 10.8300, 1.3800, 19.4300, 4.3300, 2.1100, 19.6200, -4.2300, 4.2900, 17.2100, -4.2300, 4.2900, 17.2100, -13.5500, 5.6800, 16.4400, -19.9800, 3.3700, 10.1400, -19.9800, 3.0800, 0.5100, -19.9800, 4.1200, -17.0600, -19.9800, 8.4400, -19.9700, -19.9800, 10.6300, -19.6100, -19.9800, 10.6300, -19.6100, -19.9800, 5.5000, -9.4300, -19.9800, 1.9500, -3.7400, -19.9800, -1.8000, 4.2200, -19.9800, -4.5600, 13.1500, -19.1300, -4.0500, 19.9800, -19.1300, -4.0500, 19.9800, -5.3800, -3.8200, 19.9800, 15.9100, -7.5500, 19.9800, 19.9700, -12.1300, 19.9800, 19.9700, -10.4600, 19.9800, 19.9700, -9.2700, 19.9800, 19.9700, -6.1800, 19.9800, 19.9700, -6.1800, 19.9800, 19.9700, -4.6800, 19.5600, 15.6500, -3.6800, 12.9200, 8.4000, -5.7900, 9.6900, 3.8200, -5.5500, 10.9100, -3.9100, -4.0800, 6.0200, -12.0400, -2.9300, 1.6000, -12.0400, -2.9300, 1.6000, -17.2900, -0.8900, -3.1300, -19.9800, 0.7700, -7.0300, -19.9800, 1.9800, -11.8500, -19.9800, 0.8300, -18.0000, -19.9800, -1.3800, -11.1400, -19.9800, 2.2700, -1.8800, -19.9800, 2.2700, -1.8800, -19.9800, 5.4800, 9.4500, -15.5700, 8.2800, 17.1300, -8.2900, 9.7100, 19.9700, -4.7900, 9.4900, 19.9800, 1.6000, 10.4000, 19.9800, 1.6000, 10.4000, 19.9800, 16.2600, 6.8500, 19.9800
};

int raw_feature_get_data(size_t offset, size_t length, float *out_ptr) {
    memcpy(out_ptr, features + offset, length * sizeof(float));
    return 0;
}

void setup()
{
    // put your setup code here, to run once:
    Serial.begin(115200);

    Serial.println("Edge Impulse Inferencing Demo");
}

void loop()
{
    ei_printf("Edge Impulse standalone inferencing (Arduino)\n");

    if (sizeof(features) / sizeof(float) != EI_CLASSIFIER_DSP_INPUT_FRAME_SIZE) {
        ei_printf("The size of your 'features' array is not correct. Expected %lu items, but had %lu\n",
            EI_CLASSIFIER_DSP_INPUT_FRAME_SIZE, sizeof(features) / sizeof(float));
        delay(1000);
        return;
    }

    ei_impulse_result_t result = { 0 };

    // the features are stored into flash, and we don't want to load everything into RAM
    signal_t features_signal;
    features_signal.total_length = sizeof(features) / sizeof(features[0]);
    features_signal.get_data = &raw_feature_get_data;

    // invoke the impulse
    EI_IMPULSE_ERROR res = run_classifier(&features_signal, &result, false /* debug */);
    ei_printf("run_classifier returned: %d\n", res);

    if (res != 0) return;

    ei_printf("Predictions (DSP: %d ms., Classification: %d ms., Anomaly: %d ms.): \n",
        result.timing.dsp, result.timing.classification, result.timing.anomaly);

    // print the predictions
    ei_printf("[");
    for (size_t ix = 0; ix < EI_CLASSIFIER_LABEL_COUNT; ix++) {
        ei_printf("%.5f", result.classification[ix].value);
#if EI_CLASSIFIER_HAS_ANOMALY == 1
        ei_printf(", ");
#else
        if (ix != EI_CLASSIFIER_LABEL_COUNT - 1) {
            ei_printf(", ");
        }
#endif
    }
#if EI_CLASSIFIER_HAS_ANOMALY == 1
    ei_printf("%.3f", result.anomaly);
#endif
    ei_printf("]\n");

    delay(1000);
}

void ei_printf(const char *format, ...) {
    static char print_buf[1024] = { 0 };

    va_list args;
    va_start(args, format);
    int r = vsnprintf(print_buf, sizeof(print_buf), format, args);
    va_end(args);

    if (r > 0) {
        Serial.write(print_buf);
    }
}
```

When you run this the board will hardfault. If you open the library (in `~/Documents/Arduino/library`) and edit the `tflite-trained.h` file and remove the `const` declaration on `trained_tflite` this will run fine.

We're using a very simple network (just four operations: ['DEQUANTIZE', 'FULLY_CONNECTED', 'SOFTMAX', 'QUANTIZE']). I've attached the quantized tflite model and the TF SavedModel in the library ZIP file.

## Attached

* [Arduino library, tflite file and TensorFlow SavedModel file](https://github.com/tensorflow/tensorflow/files/4902006/ei-continuous-gestures-arduino-1.0.23.zip)

**Comment by njeffrie** [comment]
Using a flatbuffer stored in flash should be fine, since we never write back to the model itself. I've ordered a Nano 33 to try to reproduce the issue you're seeing, but in the meantime, can you try the following:
-Do other (const) models (for example hello world, person detection) work on the Nano 33?
-Can you run your model on a linux machine or other MCU to see if it's something board-specific related to your model?

**Comment by janjongboom** [comment]
@njeffrie, no, this typically does not happen. We run our models on a wide variety of targets, and we have only seen this on the Arduino target.

**Comment by njeffrie** [comment]
Interesting... I'll take a look into this issue this week. I'm wondering if for some reason the flatbuffer in flash is being corrupted by something related to the BLE library. One offhand thought is that the BLE library may implicitly reserve certain addresses for BLE registers and DMA, and these could potentially conflict with the addresses used to access flash. In this case, updating the memory map could help. I highly doubt a compiler flag would cause reading from flash to fail, but I'm not very familiar with this specific setup. I'll update when I take a closer look using my Nano 33.

**Comment by janjongboom** [comment]
@njeffrie Did you ever manage to dive into this? Let me know if you need any help on our side.

**Comment by njeffrie** [comment]
Apologies for the delay. I have not gotten a chance to reproduce this issue yet. We will look into how we configure the magic wand BLE demo and compare with your code sample, since it seems like getting debug info off the board will be challenging with the hard fault.

**Comment by sushreebarsa** [comment]
This is a micro related issue,Please post this in [micro repository](https://github.com/tensorflow/tflite-micro/issues/new) and move this issue to closed status .Thanks!
